File: app/src/main/java/com/tortel/deploytrack/data/Deployment.java
/*
 * Copyright (C) 2013-2016 Scott Warner
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tortel.deploytrack.data;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.UUID;

import org.joda.time.DateTime;
import org.joda.time.Days;

import android.annotation.SuppressLint;

import com.google.firebase.database.Exclude;
import com.j256.ormlite.field.DatabaseField;
import com.j256.ormlite.table.DatabaseTable;

@DatabaseTable
public class Deployment {
	/**
	 * Use a circle-style display
	 */
	public static final int DISPLAY_CIRCLE = 0;
	/**
	 * Use a bar-style display
	 */
	public static final int DISPLAY_BAR = 1;

	private static SimpleDateFormat format;
	
	@DatabaseField(generatedId = true)
	private int id;
	@DatabaseField
	private String name;
	@DatabaseField
	private Date startDate;
	@DatabaseField
	private Date endDate;
	@DatabaseField
	private int completedColor;
	@DatabaseField
	private int remainingColor;
	@DatabaseField
	private int displayType;
    @DatabaseField
    private UUID uuid;

	@Exclude
	public String getFormattedStart(){
		if(format == null){
			getFormat();
		}
		return format.format(startDate);
	}

	@Exclude
	public String getFormattedEnd(){
		if(format == null){
			getFormat();
		}
		return format.format(endDate);
	}
	
	@SuppressLint("SimpleDateFormat")
	private void getFormat(){
		format = new SimpleDateFormat("MMM dd, yyyy");
	}

	@Exclude
	public DateTime getStart(){
		return new DateTime(startDate);
	}

	@Exclude
	public DateTime getEnd(){
		return new DateTime(endDate);
	}
	
	/**
	 * Returns the length of the deployment, in days
	 * @return the length
	 */
	public int getLength(){
		return Days.daysBetween(getStart(), getEnd()).getDays();
	}
	
	/**
	 * Get the number of days completed so far
	 * @return
	 */
	public int getCompleted(){
		DateTime start = getStart();
		//Check if it has even started
		if(start.isAfterNow()){
			return 0;
		}
		return Math.min(
				Days.daysBetween(start, new DateTime()).getDays(),
				getLength());
	}
	
	/**
	 * Get the remaining time, in days
	 * @return
	 */
	public int getRemaining(){
		return getLength() - getCompleted();
	}
	
	/**
	 * Gets the percentage completed, as a whole number (0-100)
	 * @return
	 */
	public int getPercentage(){
		return (int) ((double) getCompleted() / (double) getLength() * 100);
	}
	
	public int getId() {
		return id;
	}
	public void setId(int id) {
		this.id = id;
	}
	public String getName() {
		return name;
	}
	public void setName(String name) {
		this.name = name;
	}
	public Date getStartDate() {
		return startDate;
	}
	public void setStartDate(Date startDate) {
		this.startDate = startDate;
	}
	public Date getEndDate() {
		return endDate;
	}
	public void setEndDate(Date endDate) {
		this.endDate = endDate;
	}
	public int getCompletedColor() {
		return completedColor;
	}
	public void setCompletedColor(int completedColor) {
		this.completedColor = completedColor;
	}
	public int getRemainingColor() {
		return remainingColor;
	}
	public void setRemainingColor(int remainingColor) {
		this.remainingColor = remainingColor;
	}
	public int getDisplayType() {
		return displayType;
	}
	public void setDisplayType(int displayType) {
		this.displayType = displayType;
	}

    public UUID getUuid() {
        return uuid;
    }

    public void setUuid(UUID uuid) {
        this.uuid = uuid;
    }
}

File: app/src/main/java/com/tortel/deploytrack/data/Deployment.kt
/*
 * Copyright (C) 2013-2023 Scott Warner
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tortel.deploytrack.data

import android.annotation.SuppressLint

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import androidx.room.TypeConverter
import androidx.room.TypeConverters

import com.google.firebase.database.Exclude

import org.joda.time.DateTime
import org.joda.time.Days

import java.text.SimpleDateFormat
import java.util.Date

@Entity
@TypeConverters(DateConverter::class)
data class Deployment(val name: String, @PrimaryKey val uuid: String, val startDate: Date, val endDate: Date,
                      val completedColor: Int, val remainingColor: Int) : Comparable<Deployment> {

    override fun compareTo(other: Deployment): Int {
        /*
         * Compare start dates, and if they are the same, use end dates
         */
        val startCompare = startDate.compareTo(other.startDate)
        return if (startCompare == 0) {
            endDate.compareTo(other.endDate)
        } else startCompare
    }

    override fun toString(): String {
        return "[Deployment $uuid $name]"
    }

    @get:Exclude
    @get:Ignore
    val formattedStart: String
        get() {
            return format.format(startDate)
        }

    @get:Exclude
    @get:Ignore
    val formattedEnd: String
        get() {
            return format.format(endDate)
        }

    @get:Exclude
    @get:Ignore
    val start: DateTime
        get() = DateTime(startDate)

    @get:Exclude
    @get:Ignore
    val end: DateTime
        get() = DateTime(endDate)

    /**
     * Returns the length of the deployment, in days
     *
     * @return the length
     */
    @get:Exclude
    @get:Ignore
    val length: Int
        get() = Days.daysBetween(start, end).days

    /**
     * Get the number of days completed so far
     */
    @get:Exclude
    @get:Ignore
    val completed: Int
        get() {
            val start = start
            // Check if it has even started
            return if (start.isAfterNow) {
                0
            } else Days.daysBetween(start, DateTime()).days.coerceAtMost(length)
        }

    /**
     * Get the remaining time, in days
     */
    @get:Exclude
    @get:Ignore
    val remaining: Int
        get() = length - completed

    /**
     * Gets the percentage completed, as a whole number (0-100)
     */
    @get:Exclude
    @get:Ignore
    val percentage: Int
        get() = (completed.toDouble() / length.toDouble() * 100).toInt()


    override fun equals(other: Any?): Boolean {
        if (other is Deployment) {
            return uuid == other.uuid
        }
        return false
    }

    override fun hashCode(): Int {
        return uuid.hashCode()
    }

    companion object {
        @SuppressLint("SimpleDateFormat")
        private val format: SimpleDateFormat = SimpleDateFormat("MMM dd, yyyy")
    }
}

internal object DateConverter {
    @JvmStatic
    @TypeConverter
    fun toDate(dateLong: Long?): Date? {
        return if (dateLong == null) null else Date(dateLong)
    }

    @JvmStatic
    @TypeConverter
    fun fromDate(date: Date?): Long? {
        return date?.time
    }
}
